elastic/kibana
# [Dashboard] [Controls] Improve async controls flyout behaviour

Closes https://github.com/elastic/kibana/issues/153572

## Summary

This PR makes the calculation of the controls field registry **truly async** so that it can complete much faster - it also adds a better loading state to the field picker using the `useAsync` hook as a way to determine when the promise for the field registry has actually returned.


Previously, even though the function was marked as async and awaited, all of the internal logic was synchronous which meant that it always took `O(3 * n)` (where `n` is the number of fields, and `3` comes from the number of factories that are currently available for controls). In other words, if there are `10,000` fields in a given data view, the loop to generate the field registry would require `30,000` iterations.

This was such an intensive process that, once the calculation started, **everything** was completely consumed by it and Kibana would appear "frozen" - even the animation to open the flyout couldn't be completed, which is why it appeared to hang for so long; in fact, the animation could even get stuck in a "halfway open" state if there was a slight delay before the field registry calculation started.


### Opening Flyout
    

- **Before**

   https://user-images.githubusercontent.com/8698078/228954005-343d96b6-2029-4d9a-9501-5aff409c1f1a.mov

- **After**

    https://user-images.githubusercontent.com/8698078/228954264-c1b725f2-cf93-4e8f-884d-32a1ab07d399.mov


### Changing Data View

- **Before**

   https://user-images.githubusercontent.com/8698078/228955105-b0ef374c-7482-4a7d-9d4d-5de589a126ba.mov

- **After**

    https://user-images.githubusercontent.com/8698078/228955133-d253afff-0cea-472d-a844-09ac0765b487.mov



## How to Test

1. Load a large dataset with a large number of fields into Kibana - for example, you can run the following (which assumes that your Kibana instance is running on port `5601` and your ES server is running on port `9200`):<br><br>
    ```
    node scripts/es_archiver load test/functional/fixtures/es_archiver/large_fields --es-url=http://elastic:changeme@localhost:9200/ --kibana-url=http://elastic:changeme@localhost:5601/
    ```
    If you used the above technique to load the data, you'll also need to add a data view for that index.
2. Make this new  `large_fields` data view your **default** data view
3. Create a new dashboard and add a control from the new `large_fields` data view
4. The flyout should open **quickly** despite the fact that loading all the fields might take a little while

### Checklist



### For maintainers
